lib/web_ui/lib/src/engine/canvaskit/rasterizer.dart

```diff
@@ -5,59 +5,111 @@
-    );
-  }
+  /// A factory for creating overlays.
+  OverlayCanvasFactory<OverlayCanvas> get overlayFactory;
+
+  /// The scene host which this rasterizer should raster into.
+  DomElement get sceneHost => view.dom.sceneHost;
 
-  /// Creates a new frame from this rasterizer's surface, draws the given
-  /// [LayerTree] into it, and then submits the frame.
+  /// Draws the [layerTree] to the screen for the view associated with this
+  /// rasterizer.
   Future<void> draw(LayerTree layerTree) async {
     final ui.Size frameSize = view.physicalSize;
     if (frameSize.isEmpty) {
       // Available drawing area is empty. Skip drawing.
       return;
     }
 
-    _currentFrameSize = frameSize;
-    CanvasKitRenderer.instance.offscreenSurface.acquireFrame(_currentFrameSize);
-    viewEmbedder.frameSize = _currentFrameSize;
+    currentFrameSize = frameSize;
+    prepareToDraw();
+    viewEmbedder.frameSize = currentFrameSize;
     final CkPictureRecorder pictureRecorder = CkPictureRecorder();
-    pictureRecorder.beginRecording(ui.Offset.zero & _currentFrameSize);
+    pictureRecorder.beginRecording(ui.Offset.zero & currentFrameSize);
     pictureRecorder.recordingCanvas!.clear(const ui.Color(0x00000000));
     final Frame compositorFrame =
         context.acquireFrame(pictureRecorder.recordingCanvas!, viewEmbedder);
 
     compositorFrame.raster(layerTree, ignoreRasterCache: true);
 
-    sceneHost.prepend(renderCanvasFactory.baseCanvas.htmlElement);
-    await rasterizeToCanvas(renderCanvasFactory.baseCanvas,
-        <CkPicture>[pictureRecorder.endRecording()]);
+    sceneHost.prepend(overlayFactory.baseCanvas.htmlElement);
+    await rasterizeToCanvas(
+        overlayFactory.baseCanvas, <CkPicture>[pictureRecorder.endRecording()]);
 
     await viewEmbedder.submitFrame();
   }
 
-  /// Disposes of this rasterizer.
+  /// Do some initialization to prepare to draw a frame.
+  ///
+  /// For example, in the [OffscreenCanvasRasterizer], this ensures the backing
+  /// [OffscreenCanvas] is the correct size to draw the frame.
+  void prepareToDraw();
+
+  /// Rasterize the [pictures] to the given [canvas].
+  Future<void> rasterizeToCanvas(
+      OverlayCanvas canvas, List<CkPicture> pictures);
+
+  /// Get a [OverlayCanvas] to use as an overlay.
+  OverlayCanvas getOverlay() {
+    return overlayFactory.getCanvas();
+  }
+
+  /// Release the given [overlay] so it may be reused.
+  void releaseOverlay(OverlayCanvas overlay) {
+    overlayFactory.releaseCanvas(overlay);
+  }
+
+  /// Release all overlays.
+  void releaseOverlays() {
+    overlayFactory.releaseCanvases();
+  }
+
+  /// Remove all overlays that have been created from the DOM.
+  void removeOverlaysFromDom() {
+    overlayFactory.removeSurfacesFromDom();
+  }
+
+  /// Disposes this rasterizer.
   void dispose() {
     viewEmbedder.dispose();
-    renderCanvasFactory.dispose();
+    overlayFactory.dispose();
   }
 }
+
+abstract class OverlayCanvas {
+  DomElement get htmlElement;
```

Comment:
nit: HTML is already implied by the type; instead consider using the name to convey the function of the element, such as "hostElement", "canvasElement", or "rootElement".

Reply:
done